daemon/stop.go

```diff
@@ -38,52 +38,56 @@ func (daemon *Daemon) ContainerStop(name string, timeout *int) error {
 
 // containerStop sends a stop signal, waits, sends a kill signal.
 func (daemon *Daemon) containerStop(container *containerpkg.Container, seconds int) error {
+	// TODO propogate a context down to this function
+	ctx := context.TODO()
 	if !container.IsRunning() {
 		return nil
 	}
-
+	var wait time.Duration
+	if seconds >= 0 {
+		wait = time.Duration(seconds) * time.Second
+	}
+	success := func() error {
+		daemon.LogContainerEvent(container, "stop")
+		return nil
+	}
 	stopSignal := container.StopSignal()
-	// 1. Send a stop signal
-	if err := daemon.killPossiblyDeadProcess(container, stopSignal); err != nil {
-		// While normally we might "return err" here we're not going to
-		// because if we can't stop the container by this point then
-		// it's probably because it's already stopped. Meaning, between
-		// the time of the IsRunning() call above and now it stopped.
-		// Also, since the err return will be environment specific we can't
-		// look for any particular (common) error that would indicate
-		// that the process is already dead vs something else going wrong.
-		// So, instead we'll give it up to 2 more seconds to complete and if
-		// by that time the container is still running, then the error
-		// we got is probably valid and so we force kill it.
-		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
-		defer cancel()
 
-		if status := <-container.Wait(ctx, containerpkg.WaitConditionNotRunning); status.Err() != nil {
-			logrus.Infof("Container failed to stop after sending signal %d to the process, force killing", stopSignal)
-			if err := daemon.killPossiblyDeadProcess(container, 9); err != nil {
-				return err
-			}
-		}
+	// 1. Send a stop signal
+	err := daemon.killPossiblyDeadProcess(container, stopSignal)
+	if err != nil {
+		wait = 2 * time.Second
```

Comment:
We lost the comment explaining the reasoning for 2 seconds here.  Can we add something back in?  Otherwise this looks like a magic number without an explanation.

Also: this clobbers the value assigned to `wait` on line 48

Reply:
it is a bit of a magic number, I can add a comment but Im not sure I have anything great to say about it. If you look around these files and kill functions there are quite a lot places where it does "this failed but wait 2 seconds to see if the container exited anyways", so I was just following that pattern.

Reply:
Are you intentionally overwriting the value passed in as `seconds` though?

Reply:
yes it's intentionally being overwritten, I'm not sure I agree with the behavior but the intention was to preserve the existing behavior as much as possible while getting rid of the infinite hang.

If you look at the deleted code in line 58 you can see that if killPossiblyDeadProcess failed then the container would be given 2 seconds to exit, so the intention here was just to maintain that behavior.

Reply:
see comment below: https://github.com/moby/moby/pull/41586#issuecomment-786871667